alibi/explainers/cfrl_base.py

```diff
@@ -0,0 +1,714 @@
+
+        Parameters
+        ----------
+        shape
+            Shape of the tensor to be generated
+
+        Returns
+        -------
+        Normal noise with the appropriate mean, standard deviation and shape.
+        """
+        return self.mu + self.sigma * np.random.randn(*shape)
+
+    def __repr__(self):
+        return 'NormalActionNoise(mu={}, sigma={})'.format(self.mu, self.sigma)
+
+
+class ReplayBuffer(object):
+    """
+    Circular experience replay buffer for `CounterfactualRL`(DDPG). When the buffer is filled, then the oldest
+    experience is replaced by the new one (FIFO).
+    """
+
+    def __init__(self, size=1000):
+        """
+        Constructor.
+
+        Parameters
+        ----------
+        size
+            Dimension of the buffer in batch size. This that the total memory allocated is proportional with the
+            `size` * `batch_size`, where `batch_size` is inferred from the input tensors passed in the `append`
+            method.
+        """
+        self.x, self.x_cf = None, None           # buffers for the input and the counterfactuals
+        self.y_m, self.y_t = None, None          # buffers for the model's prediction and counterfactual target
+        self.z, self.z_cf_tilde = None, None     # buffers for the input embedding and noised counterfactual embedding
+        self.c = None                            # buffer for the conditional tensor
+        self.r_tilde = None                      # buffer for the noised counterfactual reward tensor
+
+        self.idx = 0                             # cursor for the buffer
+        self.len = 0                             # current length of the buffer
+        self.size = size                         # buffer's maximum capacity
+        self.batch_size = None                   # batch size (inferred during `append`)
+
+    def append(self,
+               x: np.ndarray,
+               x_cf: np.ndarray,
+               y_m: np.ndarray,
+               y_t: np.ndarray,
+               z: np.ndarray,
+               z_cf_tilde: np.ndarray,
+               c: Optional[np.ndarray],
+               r_tilde: np.ndarray,
+               **kwargs) -> None:
+        """
+        Adds experience to the replay buffer. When the buffer is filled, then the oldest experience is replaced
+        by the new one (FIFO).
+
+        Parameters
+        ----------
+        x
+            Input array.
+        x_cf
+            Counterfactual array.
+        y_m
+            Model's prediction class of x.
+        y_t
+            Counterfactual target class.
+        z
+            Input's embedding.
+        z_cf_tilde
+            Noised counterfactual embedding.
+        c
+            Conditional array.
+        r_tilde
+            Noised counterfactual reward array.
+        """
+        # Initialize the buffers.
+        if self.x is None:
+            self.batch_size = x.shape[0]
```

Comment:
During training, the batch_size is fixed by dropping the last one. Allowing the buffer size to have a flexible size can also generate warnings from TensorFlow because of the `@tf.function` decorator. This can also slow down the training. I would keep the batch size fixed during training.

Reply:
OK sounds good, I would just document the behaviour and the reasoning for the fixed size.